File: Python/programmers/etc/level3/단어_변환.py
from collections import deque

def solution(begin, target, words):
    def check(s1, s2):
        cnt = 0
        for i in range(len(s1)):
            if s1[i] != s2[i]:
                cnt += 1
        return cnt == 1
    
    visited = [0]*len(words)
    q = deque()
    q.append([begin, 0])
    while q:
        st, cnt = q.popleft()
        if st == target:
            return cnt
        for i in range(len(words)):
            if not visited[i] and check(st, words[i]):
                visited[i] = 1
                q.append([words[i], cnt+1])
    
    return 0


begin, target = "hit", "cog"
words = ["hot", "dot", "dog", "lot", "log", "cog"]
print(solution(begin, target, words)) # 4

begin, target = "hit", "cog"
words = ["hot", "dot", "dog", "lot", "log"]
print(solution(begin, target, words)) # 0

File: Java/programmers/etc/level3/단어변환.java
package CodeTest.Java.programmers.etc.level3;

import java.util.*;

public class 단어변환 {
    public int solution(String begin, String target, String[] words) {
        int n = words.length;
        boolean[] visited = new boolean[n];
        Queue<Node> q = new LinkedList<>();
        q.add(new Node(begin, 0));
        
        while (!q.isEmpty()) {
            Node cur = q.poll();
            String st = cur.word; int cnt = cur.edge;
            if (st.equals(target)) return cnt;
            for (int i=0; i<n; i++) {
                if (!visited[i] && check(st, words[i])) {
                    visited[i] = true;
                    q.add(new Node(words[i], cnt+1));
                }
            }
        }
        
        return 0;
    }
    
    private static boolean check(String s1, String s2) {
        int cnt = 0;
        for (int i=0; i<s1.length(); i++) {
            if (s1.charAt(i) != s2.charAt(i)) cnt++;
        }
        return cnt == 1;
    }
    
    class Node {
        String word;
        int edge;
        
        Node(String word, int edge) {
            this.word = word;
            this.edge = edge;
        }
    }
}
